vowpalwabbit/core/src/reductions/cb/cb_algs.cc

```diff
@@ -134,96 +134,110 @@ void eval_finish_example(VW::workspace& all, cb& c, VW::example& ec)
 
-base_learner* VW::reductions::cb_algs_setup(VW::setup_base_i& stack_builder)
+struct cb_algs_options_instance_v1
 {
-  options_i& options = *stack_builder.get_options();
-  VW::workspace& all = *stack_builder.get_all_pointer();
-  auto data = VW::make_unique<cb>(all.logger);
+  uint32_t num_actions;
   std::string type_string = "dr";
   bool eval = false;
   bool force_legacy = true;
+};
 
+std::unique_ptr<cb_algs_options_instance_v1> get_cb_algs_options_instance(const VW::workspace&, options_i& options)
+{
+  auto cb_algs_opts = VW::make_unique<cb_algs_options_instance_v1>();
   option_group_definition new_options("[Reduction] Contextual Bandit");
   new_options
-      .add(make_option("cb", data->cbcs.num_actions)
+      .add(make_option("cb", cb_algs_opts->num_actions)
                .keep()
                .necessary()
                .help("Use contextual bandit learning with <k> costs"))
-      .add(make_option("cb_type", type_string)
+      .add(make_option("cb_type", cb_algs_opts->type_string)
                .keep()
                .default_value("dr")
                .one_of({"ips", "dm", "dr", "mtr", "sm"})
                .help("Contextual bandit method to use"))
-      .add(make_option("eval", eval).help("Evaluate a policy rather than optimizing"))
-      .add(make_option("cb_force_legacy", force_legacy)
+      .add(make_option("eval", cb_algs_opts->eval).help("Evaluate a policy rather than optimizing"))
+      .add(make_option("cb_force_legacy", cb_algs_opts->force_legacy)
                .keep()
                .help("Default to non-adf cb implementation (cb_to_cb_adf)"));
 
   if (!options.add_parse_and_check_necessary(new_options)) { return nullptr; }
 
-  if (!eval && !force_legacy) { return nullptr; }
+  if (!cb_algs_opts->eval && !cb_algs_opts->force_legacy) { return nullptr; }
 
   // Ensure serialization of this option in all cases.
   if (!options.was_supplied("cb_type"))
   {
-    options.insert("cb_type", type_string);
+    options.insert("cb_type", cb_algs_opts->type_string);
     options.add_and_parse(new_options);
   }
 
-  cb_to_cs& c = data->cbcs;
+  if (!options.was_supplied("csoaa"))
+  {
+    std::stringstream ss;
+    ss << cb_algs_opts->num_actions;
+    options.insert("csoaa", ss.str());
+  }
+  return cb_algs_opts;
+}
+
+}  // namespace
+
+base_learner* VW::reductions::cb_algs_setup(VW::setup_base_i& stack_builder)
+{
+  options_i& options = *stack_builder.get_options();
+  VW::workspace& all = *stack_builder.get_all_pointer();
+  auto cb_algs_opts = get_cb_algs_options_instance(all, options);
+  if (cb_algs_opts == nullptr) { return nullptr; }
+  auto cb_algs_data = VW::make_unique<cb>(all.logger);
+  cb_algs_data->cbcs.num_actions = cb_algs_opts->num_actions;
+
+  cb_to_cs& c = cb_algs_data->cbcs;
 
   size_t problem_multiplier = 2;  // default for DR
-  c.cb_type = VW::cb_type_from_string(type_string);
+  c.cb_type = VW::cb_type_from_string(cb_algs_opts->type_string);
   switch (c.cb_type)
```

Comment:
Wondering if there should be a cb_type(not serialized?) and problem_multiplier(not serialized?) in cb_alg_options and the error checking below should live in get_cb_algs_options_instance ?

Reply:
moved out most of this other than the logging